### ISSUE TITLE
Incorrect parsing

### ISSUE THREAD

**Issue by kevinresol** [HaxeFoundation/haxe#5898]
The following code parses but should not.

```haxe

class Main {
	static function main() {
		$type(Extern.foo());
	}
}

extern class Extern {
	static function foo():Float-Float; // note the typo, should be Float->Float
}
```

**Comment by back2dos** [comment]
The code should parse, because essentially it is this:

```haxe
extern class Extern {
	static function foo():Float
           -Float;
}
```

However it should not pass the typer. Apparently function bodies in extern classes get completely ignored unless the function is `inline`: http://try.haxe.org/#B09A9

**Comment by kevinresol** [comment]
Oh, yes it is. Well, what if the compiler throws something like "extern function should not have a body"?

**Comment by back2dos** [comment]
IIRC in the good old days we had that error message ;)

**Comment by nadako** [comment]
I find parsing `function f():Float-Float` as `function f():Float -Float` wrong...

**Comment by back2dos** [comment]
I don't see as what else it should be parsed. Ordinarily, it's not a problem, because `-` is not valid on any type. But of course if for some reason it never goes through the typer, the result is indeed surprising.

**Comment by Simn** [comment]
I agree this should parse, but not type. Not really sure why extern non-inline expressions are allowed at the moment. Could somebody check that with 3.2 or something?

**Comment by nadako** [comment]
I remember asking about that and the answer was something regarding easily making non-extern classes extern for modular haxe applications or something like that.

**Comment by Simn** [comment]
@ncannasse: Any opinions?

**Comment by back2dos** [comment]
> I remember asking about that and the answer was something regarding easily making non-extern classes extern for modular haxe applications or something like that.

If that is the reason, then I think it makes little sense. One can always just register a build macro per `--macro addGlobalMetaData` to accomplish exactly that behavior in any specified scope (including the global one). But code just being ignored is not exactly what one would expect.

**Comment by kevinresol** [comment]
Another side-effect is that macro function inside extern class has to be marked as `inlined` in order for it to work: https://github.com/haxe-react/hxjs-redux/blob/4236aa4/src/redux/Redux.hx#L13

which is not so intuitive.

**Comment by ncannasse** [comment]
We could forbid functions body in extern classes, unless it is `inline`. If it's a problem we could still ignore empty block for the sake of backward compatibility.

**Comment by Simn** [comment]
Related: I think we should unify `extern` on classes and fields. At the moment, the field version has different semantics than the class version and is more like a "force inline".

**Comment by Simn** [comment]
This one is pretty rough, std/cpp has a couple hundred cases that are affected. I'll work through it.

And no, allowing `{ }` wouldn't help much because most of them are `return null` or the like.

There's also some creative conditional compilation like here going on: https://github.com/HaxeFoundation/haxe/blob/development/std/js/_std/haxe/Json.hx

**Comment by Simn** [comment]
Two other problems I came across:

1. If you have an `extern class` with a macro on it, it becomes a non-extern class in macro context. If there's any function without expression in that class, it errors with `Function body required` in the macro context (`python.Syntax`).

2. Eval's Sys.hx used to reference `sys.io.FileOutput` in its `extern function stderr` expression. This was used to cause the class to load. Can be fixed by moving equivalent code to `__init__` or something.

**Comment by Simn** [comment]
I've made it a warning for now because some cases are mildly annoying to work around.
